File: app/src/main/java/com/botpa/turbophotos/home/HomeActivity.java
package com.botpa.turbophotos.home;

import android.annotation.SuppressLint;
import android.app.ActivityOptions;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.content.res.Configuration;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.provider.Settings;
import android.util.DisplayMetrics;
import android.view.View;

import androidx.activity.EdgeToEdge;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.Manifest;
import android.widget.LinearLayout;

import com.botpa.turbophotos.backup.BackupActivity;
import com.botpa.turbophotos.album.AlbumActivity;
import com.botpa.turbophotos.gallery.GalleryActivity;
import com.botpa.turbophotos.settings.SettingsActivity;
import com.botpa.turbophotos.gallery.actions.Action;
import com.botpa.turbophotos.gallery.Album;
import com.botpa.turbophotos.gallery.Library;
import com.botpa.turbophotos.R;
import com.botpa.turbophotos.util.Orion;
import com.botpa.turbophotos.util.Storage;
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView;

import org.jetbrains.annotations.NotNull;

@SuppressLint({"SetTextI18n", "NotifyDataSetChanged"})
public class HomeActivity extends GalleryActivity {

    //Permissions
    private boolean permissionCheck = false;
    private boolean permissionWrite = false;
    private boolean permissionMedia = false;
    private boolean permissionNotifications = false;
    private final ActivityResultLauncher<String> requestPermissionLauncher = registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
        permissionNotifications = isGranted;
        checkPermissions();
    });

    //Activity
    private static boolean shouldReloadOnResume = false;
    private boolean isLoaded = false;
    private boolean skipResume = true;

    //Events
    private final Library.RefreshEvent onRefresh = this::manageRefresh;
    private final Library.ActionEvent onAction = this::manageAction;

    //Adapters
    private HomeAdapter adapter;
    private GridLayoutManager layoutManager;

    //Views (navbar)
    private View navbarBackup;
    private View navbarSettings;

    //Views (list)
    private SwipeRefreshLayout refreshLayout;
    private FastScrollRecyclerView list;

    //Views (loading indicator)
    private View loadingIndicator;

    //Views (system)
    private View systemNotificationsBar;
    private View systemNavigationBar;


    //Activity
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.home_screen);

        //Enable HDR
        getWindow().setColorMode(ActivityInfo.COLOR_MODE_HDR);

        //Load storage
        Storage.load(HomeActivity.this);

        //Load views & add listeners
        loadViews();
        addListeners();

        //Check permissions
        checkPermissions();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        //Remove events
        Library.removeOnRefreshEvent(onRefresh);
        Library.removeOnActionEvent(onAction);
    }

    @Override
    public void onResume() {
        super.onResume();

        //Skip first resume (called after onCreate)
        if (skipResume) {
            skipResume = false;
            return;
        }

        //Reload
        if (shouldReloadOnResume) {
            shouldReloadOnResume = false;
            recreate();
            return;
        }

        //Check for permissions
        if (permissionCheck) {
            permissionCheck = false;
            checkPermissions();
            return;
        }

        //Library not loaded -> Skip next
        if (!isLoaded) return;

        //Update horizontal item count
        updateHorizontalItemCount();

        //Refresh albums (check for new items)
        Library.loadLibrary(HomeActivity.this, false);
    }

    @Override
    public void onConfigurationChanged(@NotNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);

        //Update horizontal item count
        updateHorizontalItemCount();
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        checkPermissions();
    }

    private void checkPermissions() {
        //Update permissions
        if (Environment.isExternalStorageManager()) {
            permissionWrite = true;
            findViewById(R.id.permissionWrite).setAlpha(0.5f);
        }
        if (ContextCompat.checkSelfPermission(HomeActivity.this, Manifest.permission.READ_MEDIA_IMAGES) == PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(HomeActivity.this, Manifest.permission.READ_MEDIA_VIDEO) == PackageManager.PERMISSION_GRANTED) {
            permissionMedia = true;
            findViewById(R.id.permissionMedia).setAlpha(0.5f);
        }
        if (NotificationManagerCompat.from(HomeActivity.this).areNotificationsEnabled()) {
            permissionNotifications = true;
            findViewById(R.id.permissionNotifications).setAlpha(0.5f);
        }

        //Check if permissions are granted
        if (permissionWrite && permissionMedia && permissionNotifications) {
            //Hide permission layout
            findViewById(R.id.permissionLayout).setVisibility(View.GONE);

            //Init activity
            initActivity();
        } else {
            //Show permission layout
            findViewById(R.id.permissionLayout).setVisibility(View.VISIBLE);

            //Add request permission button listeners
            findViewById(R.id.permissionWrite).setOnClickListener(view -> {
                //Already has permission
                if (permissionWrite) return;

                //Ask for permission
                permissionCheck = true;
                Intent intent = new Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
                intent.setData(Uri.fromParts("package", getPackageName(), null));
                startActivity(intent);
            });

            findViewById(R.id.permissionMedia).setOnClickListener(view -> {
                //Already has permission
                if (permissionMedia) return;

                //Ask for permission
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    ActivityCompat.requestPermissions(this, new String[] { Manifest.permission.READ_MEDIA_IMAGES, Manifest.permission.READ_MEDIA_VIDEO }, 0);
                } else {
                    ActivityCompat.requestPermissions(this, new String[] { Manifest.permission.READ_EXTERNAL_STORAGE }, 0);
                }
            });

            findViewById(R.id.permissionNotifications).setOnClickListener(view -> {
                //Already has permission
                if (permissionNotifications) return;

                //Ask for permission
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    requestPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS);
                }
            });
        }
    }

    private void initActivity() {
        //Init adapters
        initAdapters();

        //Hide list
        list.setVisibility(View.GONE);

        //Load albums
        new Thread(() -> {
            //Load albums
            Library.loadLibrary(HomeActivity.this, true);

            //Show list
            runOnUiThread(() -> {
                //Hide loading indicator
                loadingIndicator.setVisibility(View.GONE);

                //Show list
                Orion.showAnim(list);

                //Reload albums list
                adapter.notifyDataSetChanged();
            });

            //Add events
            Library.addOnRefreshEvent(onRefresh);
            Library.addOnActionEvent(onAction);

            //Mark as loaded
            isLoaded = true;
        }).start();
    }

    public static void reloadOnResume() {
        //Reload on resume
        shouldReloadOnResume = true;
    }

    //Views
    private void loadViews() {
        //Navbar
        navbarBackup = findViewById(R.id.navbarBackup);
        navbarSettings = findViewById(R.id.navbarSettings);

        //List
        refreshLayout = findViewById(R.id.refreshLayout);
        list = findViewById(R.id.list);

        //Loading indicator
        loadingIndicator = findViewById(R.id.loadingIndicator);

        //System
        systemNotificationsBar = findViewById(R.id.notificationsBar);
        systemNavigationBar = findViewById(R.id.navigationBar);


        //Insets (keyboard & system bars)
        Orion.addInsetsChangedListener(
                findViewById(R.id.content),
                new int[] {
                        WindowInsetsCompat.Type.systemBars()
                },
                (view, insets, duration) -> {
                    refreshLayout.setProgressViewOffset(false, 0, insets.top + 50);
                    list.setPadding(0, insets.top, 0, list.getPaddingBottom() + insets.bottom);
                }
        );
        Orion.addInsetsChangedListener(
                findViewById(R.id.layout),
                new int[] {
                        WindowInsetsCompat.Type.ime(),
                        WindowInsetsCompat.Type.systemBars()
                },
                200,
                (view, insets, percent) -> {
                    //Check if keyboard is open
                    WindowInsetsCompat windowInsets = ViewCompat.getRootWindowInsets(view);
                    if (windowInsets != null && windowInsets.isVisible(WindowInsetsCompat.Type.ime())) {
                        //Keyboard is open -> Only use keyboard insets
                        insets = windowInsets.getInsets(WindowInsetsCompat.Type.ime());
                    }

                    //Update insets
                    Orion.onInsetsChangedDefault.run(view, insets, percent);
                }
        );

        //Insets (system bars background)
        Orion.addInsetsChangedListener(
                systemNotificationsBar,
                new int[] {
                        WindowInsetsCompat.Type.systemBars()
                },
                (view, insets, duration) -> {
                    systemNotificationsBar.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, insets.top));
                    systemNavigationBar.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, insets.bottom));
                }
        );
    }

    private void addListeners() {
        //Navbar
        navbarSettings.setOnClickListener(view -> {
            //Open settings
            startActivity(new Intent(HomeActivity.this, SettingsActivity.class));
        });

        navbarBackup.setOnClickListener(view -> {
            //Open backup
            startActivity(new Intent(HomeActivity.this, BackupActivity.class));
        });

        //List
        refreshLayout.setOnRefreshListener(() -> {
            //Reload library
            Library.loadLibrary(HomeActivity.this, true); //Fully refresh

            //Stop refreshing
            refreshLayout.setRefreshing(false);
        });
    }

    //Events
    private void manageRefresh(boolean updated) {
        //Refresh list
        if (updated) adapter.notifyDataSetChanged();
    }

    private void manageAction(Action action) {
        //No action
        if (action.isOfType(Action.TYPE_NONE)) return;

        //Check if albums list was changed
        if (action.hasSortedAlbumsList) {
            //Sorted albums list -> Notify all
            adapter.notifyDataSetChanged();
        } else {
            //Check if trash was added, removed or updated
            switch (action.trashAction) {
                case Action.TRASH_ADDED:
                    adapter.notifyItemInserted(0);
                    break;
                case Action.TRASH_REMOVED:
                    adapter.notifyItemRemoved(0);
                    break;
                case Action.TRASH_UPDATED:
                    adapter.notifyItemChanged(0);
                    break;
            }

            //Check if albums were deleted
            if (!action.removedIndexesInAlbums.isEmpty()) {
                //Albums were deleted -> Notify items removed
                for (int albumIndex : action.removedIndexesInAlbums) {
                    //Notify position removed
                    adapter.notifyItemRemoved(adapter.getPositionFromIndex(albumIndex));
                }
            }

            //Check if albums were sorted
            if (!action.modifiedAlbums.isEmpty()) {
                //Albums were sorted -> Notify items changed
                for (Album album : action.modifiedAlbums) {
                    //Get album index
                    int albumIndex = adapter.getIndexFromAlbum(album);
                    if (albumIndex < 0 && !album.isEspecial()) continue;

                    //Notify position changed
                    adapter.notifyItemChanged(adapter.getPositionFromIndex(albumIndex));
                }
            }
        }
    }

    //Home
    private int getHorizontalItemCount() {
        boolean isHorizontal = getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE;
        DisplayMetrics metrics = getResources().getDisplayMetrics();
        float ratio = ((float) metrics.widthPixels / (float) metrics.heightPixels);

        //Get size for portrait
        int size = Storage.getInt("Settings.homeItemsPerRow", 2);

        //Return size for current orientation
        return isHorizontal ? (int) (size * ratio) : size;
    }

    private void updateHorizontalItemCount() {
        int newHorizontalItemCount = getHorizontalItemCount();
        if (layoutManager.getSpanCount() != newHorizontalItemCount) layoutManager.setSpanCount(newHorizontalItemCount);
    }

    private void initAdapters() {
        //Create layout manager
        layoutManager = new GridLayoutManager(HomeActivity.this, getHorizontalItemCount());
        list.setLayoutManager(layoutManager);

        //Init home adapter
        adapter = new HomeAdapter(HomeActivity.this, Library.albums);
        adapter.setOnClickListener((view, album) -> {
            //Create open animation
            int startX = view.getLeft() + (view.getWidth() / 2);
            int startY = view.getTop() + (view.getHeight() / 2);
            ActivityOptions options = ActivityOptions.makeScaleUpAnimation(
                    list, //The view to scale from
                    startX, startY, //Starting point
                    0, 0 //Starting size
            );

            //Open album
            Intent intent = new Intent(HomeActivity.this, AlbumActivity.class);
            if (album == Library.trash) {
                intent.putExtra("albumName", "trash");
            } else if (album == Library.all) {
                intent.putExtra("albumName", "all");
            } else {
                intent.putExtra("albumIndex", Library.albums.indexOf(album));
            }
            startActivity(intent, options.toBundle());
        });
        list.setAdapter(adapter);
    }

}

File: app/src/main/java/com/botpa/turbophotos/screens/home/HomeActivity.kt
package com.botpa.turbophotos.screens.home

import android.Manifest
import android.annotation.SuppressLint
import android.app.ActivityOptions
import android.content.Intent
import android.content.res.Configuration
import android.net.Uri
import android.os.Build
import android.provider.Settings
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import androidx.core.graphics.Insets
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.botpa.turbophotos.R
import com.botpa.turbophotos.gallery.Album
import com.botpa.turbophotos.gallery.BaseActivity
import com.botpa.turbophotos.gallery.Library
import com.botpa.turbophotos.gallery.Library.ActionEvent
import com.botpa.turbophotos.gallery.Library.RefreshEvent
import com.botpa.turbophotos.gallery.StoragePairs
import com.botpa.turbophotos.gallery.actions.Action
import com.botpa.turbophotos.gallery.fastscroller.FastScroller
import com.botpa.turbophotos.gallery.fastscroller.FastScrollerBuilder
import com.botpa.turbophotos.gallery.options.OptionsGroup
import com.botpa.turbophotos.gallery.options.OptionsItem
import com.botpa.turbophotos.gallery.options.OptionsManager
import com.botpa.turbophotos.gallery.permissions.PermissionType
import com.botpa.turbophotos.gallery.views.GridListSeparator
import com.botpa.turbophotos.screens.album.AlbumActivity
import com.botpa.turbophotos.screens.home.filters.Filter
import com.botpa.turbophotos.screens.home.filters.FiltersDialog
import com.botpa.turbophotos.screens.settings.SettingsActivity
import com.botpa.turbophotos.screens.sync.SyncActivity
import com.botpa.turbophotos.util.Orion
import com.botpa.turbophotos.util.Storage

@SuppressLint("SetTextI18n", "NotifyDataSetChanged")
class HomeActivity : BaseActivity() {

     /*$   /$$
    | $$  | $$
    | $$  | $$  /$$$$$$  /$$$$$$/$$$$   /$$$$$$
    | $$$$$$$$ /$$__  $$| $$_  $$_  $$ /$$__  $$
    | $$__  $$| $$  \ $$| $$ \ $$ \ $$| $$$$$$$$
    | $$  | $$| $$  | $$| $$ | $$ | $$| $$_____/
    | $$  | $$|  $$$$$$/| $$ | $$ | $$|  $$$$$$$
    |__/  |__/ \______/ |__/ |__/ |__/ \______*/

    //Activity
    override val permissions: List<PermissionType> = listOf(PermissionType.Storage, PermissionType.Media)
    override val contentViewResource: Int = R.layout.home_screen

    private var isLibraryLoaded = false
    private var isLibraryLoading = false
    private var isInit = false

    //Events
    private val onRefresh = RefreshEvent { updated -> this.manageRefresh(updated) }
    private val onAction = ActionEvent { action -> this.manageAction(action) }

    //Item picker for external apps
    private var isPicking = false //An app requested to pick an item
    private val pickerLauncher = registerForActivityResult(StartActivityForResult()) { result: ActivityResult ->
        //Check if result is valid
        if (result.resultCode != RESULT_OK || result.data == null) return@registerForActivityResult

        //Return result
        val data: Intent = result.data!!
        setResult(RESULT_OK, data)
        finish()
    }

    //Permissions
    private val requestPermissionMedia = registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { granted: Map<String, Boolean> ->
        permissionManager.notifyPermissionChanged(PermissionType.Media)
        checkPermissions()
    }

    private lateinit var permissionLayout: View
    private lateinit var permissionStorage: View
    private lateinit var permissionMedia: View

    //List
    private lateinit var homeLayoutManager: GridLayoutManager
    private lateinit var homeDecorator: GridListSeparator
    private lateinit var homeAdapter: HomeAdapter

    private lateinit var homeRefreshLayout: SwipeRefreshLayout
    private lateinit var homeList: RecyclerView
    private lateinit var homeFastScroller: FastScroller

      /*$$$$$              /$$     /$$
     /$$__  $$            | $$    |__/
    | $$  \ $$  /$$$$$$  /$$$$$$   /$$  /$$$$$$  /$$$$$$$   /$$$$$$$
    | $$  | $$ /$$__  $$|_  $$_/  | $$ /$$__  $$| $$__  $$ /$$_____/
    | $$  | $$| $$  \ $$  | $$    | $$| $$  \ $$| $$  \ $$|  $$$$$$
    | $$  | $$| $$  | $$  | $$ /$$| $$| $$  | $$| $$  | $$ \____  $$
    |  $$$$$$/| $$$$$$$/  |  $$$$/| $$|  $$$$$$/| $$  | $$ /$$$$$$$/
     \______/ | $$____/    \___/  |__/ \______/ |__/  |__/|_______/
              | $$
              | $$
              |_*/

    private val options: MutableList<OptionsGroup> = ArrayList()
    private lateinit var optionsManager: OptionsManager

    private lateinit var optionSync: OptionsItem
    private lateinit var optionSettings: OptionsItem
    private lateinit var optionFilters: OptionsItem

      /*$$$$$    /$$     /$$
     /$$__  $$  | $$    | $$
    | $$  \ $$ /$$$$$$  | $$$$$$$   /$$$$$$   /$$$$$$
    | $$  | $$|_  $$_/  | $$__  $$ /$$__  $$ /$$__  $$
    | $$  | $$  | $$    | $$  \ $$| $$$$$$$$| $$  \__/
    | $$  | $$  | $$ /$$| $$  | $$| $$_____/| $$
    |  $$$$$$/  |  $$$$/| $$  | $$|  $$$$$$$| $$
     \______/    \___/  |__/  |__/ \_______/|_*/

    //Views (system)
    private lateinit var systemNotificationsBar: View
    private lateinit var systemNavigationBar: View

    //Views (navbar)
    private lateinit var navbarSubtitle: TextView
    private lateinit var navbarOptions: View

    //Views (loading indicator)
    private lateinit var loadingIndicator: View



     /*$   /$$
    | $$  | $$
    | $$  | $$  /$$$$$$  /$$$$$$/$$$$   /$$$$$$
    | $$$$$$$$ /$$__  $$| $$_  $$_  $$ /$$__  $$
    | $$__  $$| $$  \ $$| $$ \ $$ \ $$| $$$$$$$$
    | $$  | $$| $$  | $$| $$ | $$ | $$| $$_____/
    | $$  | $$|  $$$$$$/| $$ | $$ | $$|  $$$$$$$
    |__/  |__/ \______/ |__/ |__/ |__/ \______*/

    //Activity
    override fun onBeforeInitViews() {
        //Add events
        Library.addOnRefreshEvent(onRefresh)
        Library.addOnActionEvent(onAction)

        //Init options
        optionsManager = OptionsManager(this, options, backManager) { onUpdateOptions() }
    }

    override fun onInitViews() {
        //Permissions
        permissionLayout = findViewById(R.id.permissionLayout)
        permissionStorage = findViewById(R.id.permissionStorage)
        permissionMedia = findViewById(R.id.permissionMedia)

        //Navbar
        navbarSubtitle = findViewById(R.id.navbarSubtitle)
        navbarOptions = findViewById(R.id.navbarOptions)

        //List
        homeRefreshLayout = findViewById(R.id.refreshLayout)
        homeList = findViewById(R.id.list)

        //Loading indicator
        loadingIndicator = findViewById(R.id.loadingIndicator)

        //System
        systemNotificationsBar = findViewById(R.id.notificationsBar)
        systemNavigationBar = findViewById(R.id.navigationBar)


        //Insets (content)
        val listMinBottomPadding = homeList.paddingBottom
        Orion.addInsetsChangedListener(
            findViewById(R.id.content),
            intArrayOf(WindowInsetsCompat.Type.systemBars())
        ) { view: View, insets: Insets, duration: Float ->
            homeRefreshLayout.setProgressViewOffset(false, 0, insets.top + 50)
            homeList.setPadding(homeList.paddingLeft, insets.top, homeList.paddingRight, listMinBottomPadding + insets.bottom)
            homeFastScroller.setPadding(0, homeList.paddingTop, 0, homeList.paddingBottom)
        }

        //Insets (layout)
        Orion.addInsetsChangedListener(
            findViewById(R.id.layout),
            intArrayOf(WindowInsetsCompat.Type.systemBars(), WindowInsetsCompat.Type.ime()),
            200f
        ) { view: View, insets: Insets, percent: Float ->
            //Local insets var
            var insets = insets

            //Check if keyboard is open
            val windowInsets = ViewCompat.getRootWindowInsets(view)
            if (windowInsets != null && windowInsets.isVisible(WindowInsetsCompat.Type.ime())) {
                //Keyboard is open -> Only use keyboard insets
                insets = windowInsets.getInsets(WindowInsetsCompat.Type.ime())
            }

            //Update insets
            Orion.onInsetsChangedDefault.run(view, insets, percent)
        }

        //Insets (system bars background)
        Orion.addInsetsChangedListener(
            systemNotificationsBar,
            intArrayOf(WindowInsetsCompat.Type.systemBars())
        ) { view: View, insets: Insets, duration: Float ->
            systemNotificationsBar.layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, insets.top)
            systemNavigationBar.layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, insets.bottom)
        }

        //Insets (options layout)
        Orion.addInsetsChangedListener(
            optionsManager.layout,
            intArrayOf(WindowInsetsCompat.Type.systemBars())
        )
    }

    override fun onInitListeners() {
        //Navbar
        navbarOptions.setOnClickListener { view: View -> optionsManager.toggle(true) }

        //List
        homeRefreshLayout.setOnRefreshListener {
            //Reload library
            Library.loadLibrary(this, true)

            //Stop refreshing
            homeRefreshLayout.isRefreshing = false
        }

        //Options
        optionSync = OptionsItem(R.drawable.sync, "Sync") {
            //Not loaded
            if (!isLibraryLoaded) return@OptionsItem

            //Open sync
            startActivity(Intent(this, SyncActivity::class.java))
        }

        optionSettings = OptionsItem(R.drawable.settings, "Settings") {
            //Not loaded
            if (!isLibraryLoaded) return@OptionsItem

            //Open sync
            startActivity(Intent(this, SettingsActivity::class.java))
        }

        optionFilters = OptionsItem(R.drawable.filter, "Filters") {
            //Create filters list
            val filters = listOf(
                Filter(R.drawable.gallery_all, "All items", "*/*"),
                Filter(R.drawable.gallery_image, "Only images", "image/*"),
                Filter(R.drawable.gallery_video, "Only videos", "video/*")
            )

            //Create dialog
            FiltersDialog(this, filters).buildAndShow()
        }
    }

    override fun onAfterInitViews() {
        //Init components
        initHomeList()
    }

    override fun onPermissionsGranted() {
        //Hide permissions layout
        permissionLayout.visibility = View.GONE

        //Already loading or loaded
        if (isLibraryLoading || isLibraryLoaded) return
        isLibraryLoading = true

        //Hide list
        homeList.visibility = View.GONE

        //Check intent
        val filter: String
        val intent = getIntent()
        val action = intent.action
        if (action == Intent.ACTION_GET_CONTENT || action == Intent.ACTION_PICK) {
            //An app requested to select an item
            isPicking = true
            filter = intent.type ?: "*/*"
        } else {
            //Regular open
            isPicking = false
            filter = "*/*"
        }
        navbarOptions.visibility = if (isPicking) View.GONE else View.VISIBLE

        //Show loading indicator
        loadingIndicator.visibility = View.VISIBLE

        //Load library
        Thread {
            //Load library
            Library.loadLibrary(this, filter)

            //Show list
            runOnUiThread {
                //Hide loading indicator
                loadingIndicator.visibility = View.GONE

                //Show list
                Orion.animateShow(homeList)

                //Reload albums list
                homeAdapter.notifyDataSetChanged()
            }

            //Mark as loaded
            isLibraryLoaded = true
            isLibraryLoading = false
        }.start()

        //Mark as init
        isInit = true
    }

    override fun onPermissionsDenied() {
        //Show permissions layout
        permissionLayout.visibility = View.VISIBLE

        //Update buttons
        if (permissionManager.hasPermission(PermissionType.Storage)) {
            permissionStorage.isEnabled = false
        } else {
            permissionStorage.setOnClickListener { view: View ->
                //Already has permission
                if (permissionManager.hasPermission(PermissionType.Storage)) return@setOnClickListener

                //Ask for permission
                val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
                intent.data = Uri.fromParts("package", packageName, null)
                startActivity(intent)
            }
        }

        if (permissionManager.hasPermission(PermissionType.Media)) {
            permissionMedia.isEnabled = false
        } else {
            permissionMedia.setOnClickListener { view: View ->
                //Already has permission
                if (permissionManager.hasPermission(PermissionType.Media)) return@setOnClickListener

                //Ask for permission
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    requestPermissionMedia.launch(arrayOf(Manifest.permission.READ_MEDIA_IMAGES, Manifest.permission.READ_MEDIA_VIDEO))
                } else {
                    requestPermissionMedia.launch(arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE))
                }
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        //Remove events
        Library.removeOnRefreshEvent(onRefresh)
        Library.removeOnActionEvent(onAction)
    }

    override fun onResume() {
        super.onResume()

        //Not init
        if (!isInit) return

        //Check for permissions
        if (!permissionManager.hasAllPermissions) {
            if (!permissionManager.hasPermission(PermissionType.Storage)) {
                permissionManager.notifyPermissionChanged(PermissionType.Storage)
            }
            checkPermissions()
            return
        }

        //Library not loaded
        if (!isLibraryLoaded) return

        //Update list items per row
        updateListItemsPerRow()
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)

        //Update list items per row
        updateListItemsPerRow()
    }

    //Events
    private fun manageRefresh(updated: Boolean) {
        runOnUiThread {
            //Didn't update
            if (!updated) return@runOnUiThread

            //Refresh list
            homeAdapter.notifyDataSetChanged()

            //Update subtitle
            updateNavbarSubtitle()
        }
    }

    private fun manageAction(action: Action) {
        //No action
        if (action.isOfType(Action.TYPE_NONE)) return

        //Check if albums list was changed
        if (action.hasSortedAlbumsList) {
            //Sorted albums list -> Notify all
            homeAdapter.notifyDataSetChanged()
        } else {
            //Check if albums were deleted
            if (!action.removedIndexesInAlbums.isEmpty()) {
                //Albums were deleted -> Notify items removed
                for (albumIndex in action.removedIndexesInAlbums) {
                    //Notify position removed
                    homeAdapter.notifyItemRemoved(homeAdapter.getPositionFromIndex(albumIndex))
                }
            }

            //Check if albums were sorted
            var specialAlbumWasModified = false
            if (!action.modifiedAlbums.isEmpty()) {
                //Albums were sorted -> Notify items changed
                for (album in action.modifiedAlbums) {
                    //Check if album is special
                    if (album.isSpecial) {
                        specialAlbumWasModified = true
                        continue
                    }

                    //Notify album position changed
                    val albumIndex = homeAdapter.getIndexFromAlbum(album)
                    homeAdapter.notifyItemChanged(homeAdapter.getPositionFromIndex(albumIndex))
                }
            }
            if (specialAlbumWasModified) homeAdapter.notifyItemChanged(0)
        }
    }

    //Home
    private fun initHomeList() {
        //Init home layout manager
        homeLayoutManager = GridLayoutManager(this, listItemsPerRow)
        homeLayoutManager.spanSizeLookup = object : GridLayoutManager.SpanSizeLookup() {
            override fun getSpanSize(position: Int): Int {
                return if (homeAdapter.getItemViewType(position) == 0) homeLayoutManager.spanCount else 1
            }
        }
        homeList.setLayoutManager(homeLayoutManager)
        homeDecorator = GridListSeparator(20, homeLayoutManager.spanCount, 1)
        homeList.addItemDecoration(homeDecorator)

        //Init home adapter
        homeAdapter = HomeAdapter(this, Library.albums)
        homeAdapter.onClick = { view: View, album: Album ->
            //Create open animation
            val startX = view.left + (view.width / 2)
            val startY = view.top + (view.height / 2)
            val options = ActivityOptions.makeScaleUpAnimation(
                //The view to scale from
                homeList,
                //Starting point
                startX, startY,
                //Starting size
                0, 0
            )

            //Prepare intent info
            val intent = Intent(this, AlbumActivity::class.java)
            when (album) {
                Library.trash ->
                    intent.putExtra("albumName", "trash")
                Library.all ->
                    intent.putExtra("albumName", "all")
                Library.favourites ->
                    intent.putExtra("albumName", "favourites")
                else ->
                    intent.putExtra("albumIndex", Library.albums.indexOf(album))
            }

            //Open album
            if (isPicking) {
                //External item picker
                intent.putExtra("isPicking", true)
                pickerLauncher.launch(intent)
            } else {
                //Regular open
                startActivity(intent, options.toBundle())
            }
        }
        homeList.setAdapter(homeAdapter)

        //Init home fast scroller
        homeFastScroller = FastScrollerBuilder(homeList)
            .setHasHeader(true)
            .build()
    }

      /*$$$$$              /$$     /$$
     /$$__  $$            | $$    |__/
    | $$  \ $$  /$$$$$$  /$$$$$$   /$$  /$$$$$$  /$$$$$$$   /$$$$$$$
    | $$  | $$ /$$__  $$|_  $$_/  | $$ /$$__  $$| $$__  $$ /$$_____/
    | $$  | $$| $$  \ $$  | $$    | $$| $$  \ $$| $$  \ $$|  $$$$$$
    | $$  | $$| $$  | $$  | $$ /$$| $$| $$  | $$| $$  | $$ \____  $$
    |  $$$$$$/| $$$$$$$/  |  $$$$/| $$|  $$$$$$/| $$  | $$ /$$$$$$$/
     \______/ | $$____/    \___/  |__/ \______/ |__/  |__/|_______/
              | $$
              | $$
              |_*/

    private fun onUpdateOptions() {
        options.add(OptionsGroup(mutableListOf<OptionsItem>().apply {
            add(optionSync)
            add(optionSettings)
        }))
        options.add(OptionsGroup(mutableListOf<OptionsItem>().apply {
            add(optionFilters)
        }))
    }

      /*$$$$$    /$$     /$$
     /$$__  $$  | $$    | $$
    | $$  \ $$ /$$$$$$  | $$$$$$$   /$$$$$$   /$$$$$$
    | $$  | $$|_  $$_/  | $$__  $$ /$$__  $$ /$$__  $$
    | $$  | $$  | $$    | $$  \ $$| $$$$$$$$| $$  \__/
    | $$  | $$  | $$ /$$| $$  | $$| $$_____/| $$
    |  $$$$$$/  |  $$$$/| $$  | $$|  $$$$$$$| $$
     \______/    \___/  |__/  |__/ \_______/|_*/

    //List grid
    private val listItemsPerRow: Int get() {
        //Check if in horizontal orientation
        val isHorizontal = getResources().configuration.orientation == Configuration.ORIENTATION_LANDSCAPE

        //Get portrait aspect ratio
        val metrics = getResources().displayMetrics
        val ratio = (metrics.widthPixels.toFloat() / metrics.heightPixels.toFloat())

        //Get portrait items per row
        val itemsPerRow = Storage.getInt(StoragePairs.HOME_ITEMS_PER_ROW)

        //Return items per row for current orientation
        return if (isHorizontal) (itemsPerRow * ratio).toInt() else itemsPerRow
    }

    private fun updateListItemsPerRow() {
        val newItemsPerRow = listItemsPerRow
        if (homeLayoutManager.spanCount != newItemsPerRow) {
            homeLayoutManager.setSpanCount(newItemsPerRow)
            homeDecorator.spanCount = newItemsPerRow
            homeList.invalidateItemDecorations()
        }
    }

    //Navbar
    private fun updateNavbarSubtitle() {
        //Check if a filter is applied & toggle subtitle
        val filter = Library.libraryFilter
        val isFiltered = filter != "*/*"
        navbarSubtitle.visibility = if (isFiltered) View.VISIBLE else View.GONE
        if (!isFiltered) return

        //Parse filter & update subtitle
        val parts = filter.split("/")
        val type = parts[0]
        val format = parts[1]
        val subtitle = StringBuilder("Showing only ")
        when (type) {
            "image" -> subtitle.append("images ")
            "video" -> subtitle.append("videos ")
            else -> subtitle.append("custom ")
        }
        if (format != "*") subtitle.append("($format)")
        navbarSubtitle.text = subtitle.toString()
    }

}
